backstory/src/test/java/com/google/sps/story/StoryEndingToolsTest.java

```diff
@@ -0,0 +1,229 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.story;
+
+import com.google.sps.story.data.StoryEndingTools;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Quality Tests for StoryEndingTools
+ */
+@RunWith(JUnit4.class)
+public final class StoryEndingToolsTest {
+
+  // strings to use in testing story ending tools
+  private final static String EMPTY_TEXT = "";
+  private final static String SENTENCE_FRAGMENT = "This story is not ";
+  private final static String COMPLETE_SENTENCE = "This story is not incomplete.";
+  private final static String QUESTION_FRAGMENT = "What is this? I don't ";
+  private final static String COMPLETE_QUESTION = "I didn't realize. What's happening here?";
+  private final static String EXCLAMATION_FRAGMENT = "Oh no! This isn't ";
+  private final static String COMPLETE_EXCLAMATION = "I had no idea. Oh no!";
+
+  // aggregate all of these into one input array
+  private final static String[] INPUTS = { EMPTY_TEXT, SENTENCE_FRAGMENT, COMPLETE_SENTENCE, 
+                                    QUESTION_FRAGMENT, COMPLETE_QUESTION, 
+                                    EXCLAMATION_FRAGMENT, COMPLETE_EXCLAMATION };
+
+  // array of sentence enders
+  private final static String[] SENTENCE_ENDERS = {".", "?", "!"};
+
+  /**
+   * Check that sending in null input to endStory() 
+   * causes an IllegalArgumentException.
+   */
+  @Test (expected = IllegalArgumentException.class)
+  public void nullInputToEndStory() {
+    StoryEndingTools.endStory(null);
+  }
+
+  /**
+   * Check that sending in null input to 
+   * removeSentenceFragmentAtEnd() to make sure
+   * it throws an IllegalArgumentException.
+   */
+  @Test (expected = IllegalArgumentException.class)
+  public void nullInputToRemoveSentenceFragment() {
+    StoryEndingTools.removeSentenceFragmentAtEnd(null);
+  }
+
+  /**
+   * Check that sending in null input to addEnding()
+   * to make sure it throws an IllegalArgumentException.
+   */
+  @Test (expected = IllegalArgumentException.class)
+  public void nullInputToAddEnding() {
+    StoryEndingTools.addEnding(null);
+  }
+
+  /**
+   * Private helper method to check if text ends in punctuation.
+   * 
+   * @param text the text to check ending for
+   * @return true, if it's empty or if the char at the last index is in SENTENCE_ENDERS
+   *
+   */
+  private static boolean endsInSentenceEndingPunctuation(String text) {
+    if (text.equals("")) {
+      return true;
```

Comment:
This adds an element of uncertainty to the tests:

```
String sentence = foo();
Assert.assertTrue(endsInSentenceEndingPunctuation(sentence));
```

If I see this, I won't be certain that sentence indeed has punctuation.

I'd recommend either failing the test here, or removing this block, to avoid introducing additional complexity.

Reply:
I'll remove this block because I agree, it isn't relevant.

Reply:
Done